itests/gateway_test.go

```diff
@@ -297,3 +359,209 @@ func sendFunds(ctx context.Context, fromNode *kit.TestFullNode, fromAddr address
 
 	return nil
 }
+
+func TestGatewayRateLimits(t *testing.T) {
+	req := require.New(t)
+
+	kit.QuietMiningLogs()
+	ctx := context.Background()
+	tokensPerSecond := 10
+	requestsPerMinute := 30 // http requests
+	nodes := startNodes(ctx, t,
+		withNodeOpts(kit.DisableEthRPC()),
+		withPerConnectionAPIRateLimit(tokensPerSecond),
+		withPerHostRequestsPerMinute(requestsPerMinute),
+	)
+
+	time.Sleep(time.Second)
+
+	// ChainHead uses chainRateLimitTokens=2.
+	// But we're also competing with the paymentChannelSettler which listens to the chain uses
+	// ChainGetBlockMessages on each change, which also uses chainRateLimitTokens=2.
+	// So each loop should be 4 tokens.
+	loops := 10
+	tokensPerLoop := 4
+	start := time.Now()
+	for i := 0; i < loops; i++ {
+		_, err := nodes.lite.ChainHead(ctx)
+		req.NoError(err)
+	}
+	tokensUsed := loops * tokensPerLoop
+	expectedEnd := start.Add(time.Duration(float64(tokensUsed) / float64(tokensPerSecond) * float64(time.Second)))
+	allowPad := time.Duration(float64(tokensPerLoop) / float64(tokensPerSecond) * float64(time.Second)) // add padding to account for slow test runs
+	t.Logf("expected end: %s, now: %s, allowPad: %s, actual delta: %s", expectedEnd, time.Now(), allowPad, time.Since(expectedEnd))
+	req.WithinDuration(expectedEnd, time.Now(), allowPad)
+
+	client := &http.Client{}
+	url := fmt.Sprintf("http://%s/rpc/v1", nodes.gatewayAddr)
+	jsonPayload := []byte(`{"method":"Filecoin.ChainHead","params":[],"id":1,"jsonrpc":"2.0"}`)
+	var failed bool
+	for i := 0; i < requestsPerMinute*2 && !failed; i++ {
+		func() {
+			request, err := http.NewRequest("POST", url, bytes.NewBuffer(jsonPayload))
+			req.NoError(err)
+			request.Header.Set("Content-Type", "application/json")
+			response, err := client.Do(request)
+			req.NoError(err)
+			defer func() { _ = response.Body.Close() }()
+			if http.StatusOK == response.StatusCode {
+				body, err := io.ReadAll(response.Body)
+				req.NoError(err)
+				result := map[string]interface{}{}
+				req.NoError(json.Unmarshal(body, &result))
+				req.NoError(err)
+				req.NotNil(result["result"])
+				height, ok := result["result"].(map[string]interface{})["Height"].(float64)
+				req.True(ok)
+				req.Greater(int(height), 0)
+			} else {
+				req.Equal(http.StatusTooManyRequests, response.StatusCode)
+				req.LessOrEqual(i, requestsPerMinute+1)
+				failed = true
+			}
+		}()
```

Comment:
Subjective: Refactoring this block into a set of functions would make it easier to read.

Reply:
I've pulled out a client rpc call here that I can reuse across the two separate tests, let me know if you think I should go further